{{>partial_header}}

#if NET6_0_OR_GREATER
using System.Net;
using System.Runtime.Serialization;
using System.Threading.Tasks;

namespace {{packageName}}.Exceptions;
#else
using System;
using System.Net;
using System.Net.Http;
using System.Runtime.Serialization;
using System.Threading.Tasks;

namespace {{packageName}}.Exceptions
{
#endif

public class ApiException : Exception {
    /// <summary>
    /// Initializes a new instance of the <see cref="ErrorApiException"/> class.
    /// </summary>
    protected ApiException()
        : base() {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ApiException"/> class with a specified error message.
    /// </summary>
    /// <param name="message">The error message that explains the reason for the exception.</param>
    protected ApiException(string message)
        : base(message) {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ApiException"/> class with a specified error message
    /// and a reference to the inner exception that is the cause of this exception.
    /// </summary>
    /// <param name="message">The error message that explains the reason for the exception.</param>
    /// <param name="innerException">The exception that is the cause of the current exception, or a null
    /// reference if no inner exception is specified.</param>
    protected ApiException(string message, Exception innerException)
        : base(message, innerException) {
    }

    /// <inheritdoc />
    protected ApiException(SerializationInfo serializationInfo, StreamingContext streamingContext)
        : base(serializationInfo, streamingContext) {
    }

    /// <summary>
    /// Create an instance of the specific exception related to a particular response.
    /// </summary>
    /// <param name="response"><see cref="HttpResponseMessage"/></param>
    /// <param name="request"><see cref="HttpRequestMessage"/></param>
    /// <param name="apiName"></param>
    /// <returns>An instance of a <see cref="ApiException"/> subclass containing the appropriate exception for this response.</returns>
    public static async Task<ApiException> CreateSpecificExceptionAsync(HttpResponseMessage? response, HttpRequestMessage request,
        string? apiName = null) {
        var statusCode = response?.StatusCode;
        
        // Handle authentication errors
        if (statusCode == HttpStatusCode.Unauthorized || statusCode == HttpStatusCode.Forbidden) {
            return await FgaApiAuthenticationError.CreateAsync(response, request, apiName).ConfigureAwait(false);
        }
        
        // Handle validation errors
        if (statusCode == HttpStatusCode.BadRequest || 
            statusCode == HttpStatusCodeExtensions.UnprocessableEntity) {
            return await FgaApiValidationError.CreateAsync(response, request, apiName).ConfigureAwait(false);
        }
        
        // Handle not found errors
        if (statusCode == HttpStatusCode.NotFound) {
            return await FgaApiNotFoundError.CreateAsync(response, request, apiName).ConfigureAwait(false);
        }
        
        // Handle rate limit errors
        if (statusCode == HttpStatusCodeExtensions.TooManyRequests) {
            return await FgaApiRateLimitExceededError.CreateAsync(response, request, apiName).ConfigureAwait(false);
        }
        
        // Handle other errors
        if (statusCode >= HttpStatusCode.InternalServerError && statusCode != HttpStatusCode.NotImplemented) {
            return await FgaApiInternalError.CreateAsync(response, request, apiName).ConfigureAwait(false);
        }
        
        // Default case
        return await FgaApiError.CreateAsync(response, request, apiName).ConfigureAwait(false);
    }
}
#if NETSTANDARD2_0
}
#endif
